### ISSUE TITLE
Design: DNS revamp

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#6666]
There have been a bunch of discussions about DNS, and I thought it was worthwhile to collect them all.

@smarterclayton @bgrant0607 @preilly @vishh  

# Background and current state

We currently serve DNS from a pod running in the cluster.  It runs etcd, skydns, and a bridge between them called kube2sky.  The only records we really serve are A records.  We add an A record for every Service object, resolving to the portal IP for that Service.  For example, given a Service named 'mysvc' in namespace 'myns', we serve a DNS A record (a simple address) for `mysvc.myns.kubernetes.local`.

## SkyDNS schema

To implement this we configure SkyDNS, through etcd,  with a single config record per service:

```
/skydns/local/kubernetes/myns/mysvc =  '{"host":"$PORTAL_IP", "port":$SERVICE_PORT}'
```

# Motivations for change

1) We now have "headless" services - services that do not have a portal IP.
2) We now have named Service ports and multi-port Services.
3) SkyDNS supports arbitrary backend implementations, eliminating the need to run an etcd instance just for DNS.
4) Several feature requests and PRs that touch on DNS.

# Goal

To evaluate options for designs and decide on one path to implement.

# Headless services

The obvious thing to serve for a headless service is the set of pod IPs that implement the service - a plain old DNS round-robin.  The reason we haven't just done that for "classic" services is that many DNS client libraries do not respect DNS TTLs, or applications assume (intentionally or not) that they can resolve a name once and then remember that result forever.  But for headless services, we don't have much choice - we really have to do SOMETHING.

The simplest solution, then, is just to resolve service names into multiple A records, pointing to pod IPs, where we would normally resolve to a single portal IP.

## SkyDNS schema

Headless services will need one config record per service-pod.

```
/skydns/local/kubernetes/myns/mysvc/$POD_NAME =  '{"host":"$POD_IP", "port":$SERVICE_PORT}'
```

SkyDNS will aggregate the pods under `mysvc` and report them all as A records for `mysvc.myns.kubernetes.local`.  This will also have the side-effect of creating A records for each `$POD_NAME.mysvc.myns.kubernetes.local`.

There's a wrinkle, though.  Where do we get pod names and IPs?  The DNS code will have to do more than just watch Services (which is all it does today).  It will have to also watch Endpoints.  When it observes a headless Service, it will look at the Endpoints for that Service and extract the endpoint IPs.

The last question here is what to use as the pod name in the above schema?  I see two options.

1) Use the `EndpointAddress.TargetRef` field.  For example, a TargetRef that points to `{Kind: "Pod", Name: "p1"}` could become "pod-p1" or even just "p1".  This has the nice effect of producing A records that match kubernetes objects.  We have to be careful about name length (DNS says 63 characters per segment) and we have to handle dots (we allow names with dots, but they have special meaning in DNS) and we have to possibly handle collisions (if we just use Name instead of Kind-Name) and we still have to handle EndpointAdress objects without a TargetRef (it is optional).

2) Generate random names.  Instead of `mypod.mysvc.myns.kubernetes.local` we would get `ag54d.mysvc.myns.kubernetes.local`, but avoid all of the concerns above.  They don't even have to be very random - just unique within the service.

## Consistency

The followup to this idea is that we should probably offer the same expansion for Endpoints under a classic Service, to enable things like DNS-SD for Prometheus.  But where?  Unfortunately, we did not leave any room in the DNS space to extend.  For any value of `*.mysvc.myns.kubernetes.local`, SkyDNS will do the wrong thing and expose multiple A records for `mysvc.myns.kubernetes.local`.  Where this behavior helps us with headless services it hurts with regular services.

Options:

1a) Inject a level in the hierarchy.  What was `mysvc.myns.kubernetes.local` becomes `mysvc.service.myns.kubernetes.local` or `mysvc.portal.myns.kubernetes.local`.  Now we can expose the endpoints directly as `mysvc.endpoints.myns.kubernetes.local`.  We can probably hide the transition by making the DNS search path be `service.myns.kubernetes.local myns.kubernetes.local kubernetes.local`.  This allows namespace local services to be resolved with a single token. It makes cross-namespace names a bit uglier: `kubernetes.default` now becomes `kubernetes.service.default`.

1b) Inject a level in the hierarchy.  What was `mysvc.myns.kubernetes.local` becomes `mysvc.myns.service.kubernetes.local` or `mysvc.myns.portal.kubernetes.local`.  Now we can expose the endpoints directly as `mysvc.myns.endpoints.kubernetes.local`.  We can probably hide the transition by making the DNS search path be `myns.service.kubernetes.local service.kubernetes.local kubernetes.local`.  This allows namespace local services to be resolved with a single token and makes cross-namespace names the same as now: `kubernetes.default` now resolves to  `kubernetes.default.service.kubernetes.local`, while allowing `kubernetes.default.endpoints`.

2) Insert a level of name.  What was `mysvc.myns.kubernetes.local` becomes `portal.mysvc.myns.kubernetes.local`.  The leaves `endpoints.mysvc.myns.kubernetes.local` available for the endpoints A records.  Accessing the master service becomes `portal.kubernetes.default`.

3) Lobby SkyDNS for some feature to do what we need.  Other users have asked for similar things. (https://github.com/skynetservices/skydns/issues/147)

4) Lobby SkyDNS to support multiple first-level domains and distinguish pods and services at that scope.  `mysvc.myns.kubernetes.local` remains while `mysvc.myns.kube-endpoints.local` is available for the endpoints A records. (https://github.com/skynetservices/skydns/issues/47)

4) Fork SkyDNS and build our own DNS server.  The skydns codebase could largely be reused.

# Multiport services and named ports

Giving ports names pretty much begs for proper DNS SRV records (services).  we can support this for both headless and "classic" services.  We should follow the RFC for SRV records, for maximum impact.  SRV records will only be generated when a port has an explicit name.

## SkyDNS schema

We will need one config record per port.

```
/skydns/local/kubernetes/myns/mysvc/_$PORT_PROTOCOL/_$PORT_NAME/$POD_NAME '{"host":"$POD_NAME.svc.ns.kubernetes.local", "port":$PORT}'
```

This will create an SRV record for `_$PORT_NAME._$PORT_PROTOCOL.svc.ns.kubernetes.local` on `$PORT`.  For example: `_http._tcp.svc.ns.kubernetes.local`.  This is RFC compliant.

We will get the port name, protocol, and number from the Endpoints object in the case of headless Services and from the Service object in case of "classic" services.  We have the same issue with pod name as above and will use the same solution for headless Services, while classic Services will just use the literal "portal".

# SkyDNS backends

The groundwork has been laid to implement SkyDNS on top of kubernetes directly, without etcd as an intermediary.  I'd like to make this change, but only AFTER the schema changes are implemented.  Once that is done, we can consider putting the DNS serving logic into each kube-proxy, since it is already caching all the information we need to serve DNS.  This does have the downside that kube-proxy is not an HA component, whereas a DNS pod could be.

# Other open issues

This does not address nominal services.  I will update this as I process the current state of that issue.  I think this schema will address the DNS-SD (service discovery) needs of systems like Prometheus in the headless case.  In the classic Services case, we need to get raw endpoints.

# Appendix

These are two scripts I worked up to manual test the schema against running SkyDNS.  They assumes you are running on the node that is running the DNS pod.

```
#!/bin/bash

function d {
docker exec -ti $(docker ps | grep etcd | cut -f1 -d' ') "$@"
}

echo multiport headless service with DNSRR
d /etcdctl rm --recursive /skydns/local/kubernetes/ns

# Given a headless Service 'svc', with 3 Endpoints, only 2 with Podrefs, each with 2 named ports and 1 unnamed port

# Create records for the hosts, based on port names.  SkyDNS will infer a base name and produce A and SRV records.
# Generate a random name for the unnamed port
UNKPORT=unknown-port-$RANDOM
d /etcdctl set /skydns/local/kubernetes/ns/svc/pod1/$UNKPORT '{"host":"1.1.1.1", "port":9376}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/pod1/http '{"host":"1.1.1.1", "port":80}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/pod1/https '{"host":"1.1.1.1", "port":443}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/pod2/$UNKPORT '{"host":"2.2.2.2", "port":9376}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/pod2/http '{"host":"2.2.2.2", "port":80}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/pod2/https '{"host":"2.2.2.2", "port":443}'
# Generate a random name for the un-referenced pod
UNKPOD=unknown-pod-$RANDOM
d /etcdctl set /skydns/local/kubernetes/ns/svc/$UNKPOD/$UNKPORT '{"host":"3.3.3.3", "port":9376}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/$UNKPOD/http '{"host":"3.3.3.3", "port":80}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/$UNKPOD/https '{"host":"3.3.3.3", "port":443}'

# Create explicit SRV respords for the RFC-style service names.  SkyDNS does not recurse through records with leading underscores.
d /etcdctl set /skydns/local/kubernetes/ns/svc/_tcp/_http/pod1 '{"host":"pod1.svc.ns.kubernetes.local", "port":80}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/_tcp/_https/pod1 '{"host":"pod1.svc.ns.kubernetes.local", "port":443}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/_tcp/_http/pod2 '{"host":"pod2.svc.ns.kubernetes.local", "port":80}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/_tcp/_https/pod2 '{"host":"pod2.svc.ns.kubernetes.local", "port":443}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/_tcp/_http/$UNKPOD "{\"host\":\"$UNKPOD.svc.ns.kubernetes.local\", \"port\":80}"
d /etcdctl set /skydns/local/kubernetes/ns/svc/_tcp/_https/$UNKPOD "{\"host\":\"$UNKPOD.svc.ns.kubernetes.local\", \"port\":443}"

# Expect multiple A records - one for each IP above
dig @10.0.0.10 +noall +answer +additional +ignore svc.ns.kubernetes.local A
read X

# Expect one A record for each pod
dig @10.0.0.10 +noall +answer +additional +ignore pod1.svc.ns.kubernetes.local A
dig @10.0.0.10 +noall +answer +additional +ignore pod2.svc.ns.kubernetes.local A
read X

# Expect one A record for each port.pod
dig @10.0.0.10 +noall +answer +additional +ignore http.pod1.svc.ns.kubernetes.local A
dig @10.0.0.10 +noall +answer +additional +ignore https.pod1.svc.ns.kubernetes.local A
dig @10.0.0.10 +noall +answer +additional +ignore http.pod2.svc.ns.kubernetes.local A
dig @10.0.0.10 +noall +answer +additional +ignore https.pod2.svc.ns.kubernetes.local A
read X

# Expect one SRV record for each pod for each port
dig @10.0.0.10 +noall +answer +additional +ignore svc.ns.kubernetes.local SRV
read X

# Expect one SRV record for each pod for each port
dig @10.0.0.10 +noall +answer +additional +ignore _http._tcp.svc.ns.kubernetes.local SRV
dig @10.0.0.10 +noall +answer +additional +ignore _https._tcp.svc.ns.kubernetes.local SRV
read X
```

```
#!/bin/bash

function d {
docker exec -ti $(docker ps | grep etcd | cut -f1 -d' ') "$@"
}

echo multiport headless service with DNSRR
d /etcdctl rm --recursive /skydns/local/kubernetes/ns

# Given a Service 'svc' with 2 named ports and 1 unnamed port

# Create records for the hosts, based on port names.  SkyDNS will infer a base name and produce A and SRV records.
# Generate a random name for the unnamed port
UNKPORT=unknown-port-$RANDOM
d /etcdctl set /skydns/local/kubernetes/ns/svc/portal/$UNKPORT '{"host":"1.1.1.1", "port":9376}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/portal/http '{"host":"1.1.1.1", "port":80}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/portal/https '{"host":"1.1.1.1", "port":443}'

# Create explicit SRV respords for the RFC-style service names.  SkyDNS does not recurse through records with leading underscores.
d /etcdctl set /skydns/local/kubernetes/ns/svc/_tcp/_http/portal '{"host":"svc.ns.kubernetes.local", "port":80}'
d /etcdctl set /skydns/local/kubernetes/ns/svc/_tcp/_https/portal '{"host":"svc.ns.kubernetes.local", "port":443}'

# Expect one A record
dig @10.0.0.10 +noall +answer +additional +ignore svc.ns.kubernetes.local A
read X

# Expect one SRV record for each port
dig @10.0.0.10 +noall +answer +additional +ignore svc.ns.kubernetes.local SRV
read X

# Expect one SRV record for each port
dig @10.0.0.10 +noall +answer +additional +ignore _http._tcp.svc.ns.kubernetes.local SRV
dig @10.0.0.10 +noall +answer +additional +ignore _https._tcp.svc.ns.kubernetes.local SRV
read X
```

**Comment by smarterclayton** [comment]
> On Apr 9, 2015, at 8:04 PM, Tim Hockin <notifications@github.com> wrote:
> 
> There have been a bunch of discussions about DNS, and I thought it was worthwhile to collect them all.
> 
> @smarterclayton @bgrant0607 @preilly @vishh
> 
> Background and current state
> 
> We currently serve DNS from a pod running in the cluster. It runs etcd, skydns, and a bridge between them called kube2sky. The only records we really serve are A records. We add an A record for every Service object, resolving to the portal IP for that Service. For example, given a Service named 'mysvc' in namespace 'myns', we serve a DNS A record (a simple address) for mysvc.myns.kubernetes.local.
> 
> SkyDNS schema
> 
> To implement this we configure SkyDNS, through etcd, with a single config record per service:
> 
> /skydns/local/kubernetes/myns/mysvc =  '{"host":"$PORTAL_IP", "port":$SERVICE_PORT}'
> Motivations for change
> 
> 1) We now have "headless" services - services that do not have a portal IP.
> 2) We now have named Service ports and multi-port Services.
> 3) SkyDNS supports arbitrary backend implementations, eliminating the need to run an etcd instance just for DNS.
> 4) Several feature requests and PRs that touch on DNS.
> 
> Goal
> 
> To evaluate options for designs and decide on one path to implement.
> 
> Headless services
> 
> The obvious thing to serve for a headless service is the set of pod IPs that implement the service - a plain old DNS round-robin. The reason we haven't just done that for "classic" services is that many DNS client libraries do not respect DNS TTLs, or applications assume (intentionally or not) that they can resolve a name once and then remember that result forever. But for headless services, we don't have much choice - we really have to do SOMETHING.
> 
> The simplest solution, then, is just to resolve service names into multiple A records, pointing to pod IPs, where we would normally resolve to a single portal IP.
> 
> SkyDNS schema
> 
> Headless services will need one config record per service-pod.
> 
> /skydns/local/kubernetes/myns/mysvc/$POD_NAME =  '{"host":"$POD_IP", "port":$SERVICE_PORT}'
> SkyDNS will aggregate the pods under mysvc and report them all as A records for mysvc.myns.kubernetes.local. This will also have the side-effect of creating A records for each $POD_NAME.mysvc.myns.kubernetes.local.
> 
> There's a wrinkle, though. Where do we get pod names and IPs? The DNS code will have to do more than just watch Services (which is all it does today). It will have to also watch Endpoints.
> 
Technically, we don't have to watch, just query on demand, and let skydns cache the result on its own.

One downside is that reverse lookup for portal IPs requires a full cache of services... But that cache could be reused by the service ip allocator controller and kill two birds with one stone.

An advantage of moving this to the kube-proxy is that the Kube-proxy has all of this info loaded already.
> When it observes a headless Service, it will look at the Endpoints for that Service and extract the endpoint IPs.
> 
> The last question here is what to use as the pod name in the above schema? I see two options.
> 
> 1) Use the EndpointAddress.TargetRef field. For example, a TargetRef that points to {Kind: "Pod", Name: "p1"} could become "pod-p1" or even just "p1". This has the nice effect of producing A records that match kubernetes objects. We have to be careful about name length (DNS says 63 characters per segment) and we have to handle dots (we allow names with dots, but they have special meaning in DNS) and we have to possibly handle collisions (if we just use Name instead of Kind-Name) and we still have to handle EndpointAdress objects without a TargetRef (it is optional).
> 
> 2) Generate random names. Instead of mypod.mysvc.myns.kubernetes.local we would get ag54d.mysvc.myns.kubernetes.local, but avoid all of the concerns above. They don't even have to be very random - just unique within the service.
> 
> Mulitport services and named ports
> 
> Giving ports names pretty much begs for proper DNS SRV records (services). we can support this for both headless and "classic" services. We should follow the RFC for SRV records, for maximum impact.  SRV records will only be generated when a port has an explicit name.
> 
> SkyDNS schema
> 
> We will need one config record per port.
> 
> /skydns/local/kubernetes/myns/mysvc/_$PORT_PROTOCOL/_$PORT_NAME/$POD_NAME '{"host":"$POD_NAME.svc.ns.kubernetes.local", "port":$PORT}'
> This will create an SRV record for _$PORT_NAME._$PORT_PROTOCOL.svc.ns.kubernetes.local on $PORT. For example: _http._tcp.svc.ns.kubernetes.local. This is RFC compliant.
> 
> We will get the port name, protocol, and number from the Endpoints object in the case of headless Services and from the Service object in case of "classic" services.  We have the same issue with pod name as above and will use the same solution for headless Services, while classic Services will just use the literal "portal".
> 
> SkyDNS backends
> 
> The groundwork has been laid to implement SkyDNS on top of kubernetes directly, without etcd as an intermediary. I'd like to make this change, but only AFTER the schema changes are implemented. Once that is done, we can consider putting the DNS serving logic into each kube-proxy, since it is already caching all the information we need to serve DNS. This does have the downside that kube-proxy is not an HA component, whereas a DNS pod could be.
> 
Although a set of pods running on >1 host are now ha with DNS because one of them will still be able to resolve even if one proxy has problems.  Reduces as well the impact of partition - the Kube proxy continues to serve DNS 
> Other open issues
> 
> This does not address nominal services. I will update this as I process the current state of that issue. I think this schema will address the DNS-SD (service discovery) needs of systems like Prometheus.
> 
We need to consider how to reserve sufficient namespace in DNS so that future use cases can be added without overlap of the space we expose (some of the examples demonstrate this but it's not explicitly called out).

**Comment by thockin** [comment]
> One downside is that reverse lookup for portal IPs requires a full cache of services...

I am still unconvinced that reverse lookup matters at all.  For portal IPs you should never see traffic with those IPs anyway.  For headless services, a given  IP could (will) have several reverse names. 

> We need to consider how to reserve sufficient namespace in DNS so that future use cases can be added without overlap of the space we expose (some of the examples demonstrate this but it's not explicitly called out).

In my rush to finish this doc, I totally forgot to explore that.  Added a section on consistency and discussed options.

**Comment by thockin** [comment]
@alml also

**Comment by Symmetric** [comment]
This PR would allow us to integrate Kubernetes services with Calico's distributed firewall. (See https://github.com/GoogleCloudPlatform/kubernetes/issues/7245 for a bit more detail on why). Looks good!

I'd be interested to know how widespread the DNS caching/TTL issues you mention are in your experience; Mesosphere are going with SRV-ish service discovery as well, so it would be great to get a feel for how much trouble this mechanism is likely to cause.

One point to note (I'm sure you're aware, but I'll mention since it wasn't explicitly called out) is that (RFC compliant) SRV responses can only advertise a single port, so that's not an immediate fit for multi-port/protocol services. 

At the very least, clients would need to do a separate lookup for each port in the service. If those ports aren't named, then how do you distinguish the lookup for port 80 and port 443 exposed on a service? Furthermore, how do clients even know what ports a service exposes in order to make the appropriate SRV lookups? In the SIP world we solve this using NAPTR records, so you do:

```
NAPTR lookup => SRV records (one for each _service/_port on the domain)
SRV lookups => A records
```

But that makes it even less likely that clients will support the lookup correctly than just using SRV... (the Node.JS Redis client supports SRV, but not NAPTR).

Mesosphere are doing something non-standard, I believe they are returning all the ports in one SRV response (and exposing an HTTP API that can return the SRV records, too). See https://github.com/mesosphere/mesos-dns/issues/107, https://github.com/mesosphere/mesos-dns/issues/61. (I might have the wrong end of the stick there, so don't take this as gospel).

**Comment by thockin** [comment]
> I'd be interested to know how widespread the DNS caching/TTL issues you
mention are in your experience

My understanding is that a large preponderance of client DNS libraries are
broken by default, though some have flags to un-break.  It's also that many
apps are broken (do a lookup, cache result forever).

> One point to note (I'm sure you're aware, but I'll mention since it
wasn't explicitly called out) is that (RFC compliant) SRV responses can
only advertise a single port, so that's not an immediate fit for
multi-port/protocol services.

Yeah, we would have a single SRV record per named-port.  As you
say,  clients would need to do a separate lookup for each port.  Unnamed
ports would not get a meaningful record.

> how do clients even know what ports a service exposes in order to make
the appropriate SRV lookups

I don't expect clients to "go shopping".  I expect them to know what it is
they want, just not where it is.  That said, you can still do an SRV query
against the base service (in this schema) and get a multi-record response,
one for each port.  This sounds like what mesos is doing and just happens
to be the default behavior of SkyDNS.  `dig` seems to handle it just fine.

**Comment by alml** [comment]
> This PR would allow us to integrate Kubernetes services with Calico's distributed firewall. (See #7245 for a bit more detail on why). Looks good!

Are you talking about my PR? (https://github.com/GoogleCloudPlatform/kubernetes/pull/6438)
Implemented DNS schema for Pods is: <podID>[.<portName>].<svcID>.<namespace>.pod.kubernetes.local
If port name is set in the service definition, then portName is added. Otherwise it's just omitted. If multiple anonymous ports exist, only one of them is exported. Current implementation picks the last one.

The schema for services is: [.<portName>].<svcID>.<namespace>.svc.kubernetes.local
Rules about port name are the same.

This is designed to fit the following use cases:

1. Connect to any backend (load-balanced). Example nginx setup:
    proxy_pass http://http.mybackend.svc.kubernetes.local/; ("http" is the port 80 name)
    proxy_pass https://https.mybackend.svc.kubernetes.local/; ("https" is the port 443 name)
2. Connect to specific backend (when debugging stuff, for example)
    http://mybackend-ja32h.http.mybackend.pod.kubernetes.local/ (mybackend-ja32h is the pod name)
3. Scrape monitoring data from "http" port from all the backends (Prometheus setup):
    http://http.mybackend.pod.kubernetes.local:80/monitor

**Comment by smarterclayton** [comment]
I seem to find myself drawn to 1b the most.

I'm tempted to suggest we define a DNS space which while recognizable is not Kubernetes specific.  We definitely want to be one segment above .local to allow others to coexist in the local space.  I would suggest we do not want to make the local space configurable, to allow anyone who can deploy to one Kube to deploy to any other.  I also want to make it easy to run a singleton cluster on a dev laptop for testing, so we should be sensitive to implications of mDNS and local setup being easy. 

    *.cluster.local
    <name>.<namespace>.svc.cluster.local # services
    *.<reserved>.cluster.local
    
I would like to reserve a subset of the svc space so wildcards work internally.

    *.all.<name>.<namespace>.svc.cluster.local

Or

    *.<name>.<namespace>.svcs.cluster.local

The proposal for srv looks fine to me as long as we have a solution for wildcards.

**Comment by smarterclayton** [comment]
I'm not really in favor of pod DNS - I want to keep DNS tied to endpoints which provides the discovery mechanism for names.

I also think we should be exposing the primary names for a service under the service status field so that clients can discover what names are available.  We should prepare for there being multiple DNS routes to a service, some from outside the network (we plan to move routes in that direction so you can define ingress into the cluster in a separate object, and have those objects decorate the service to announce their presence.

**Comment by dcieslak19973** [comment]
Would it make sense to think about services like ZooKeeper where it might be useful to be able to refer to a specific instance in a pod? Perhaps its the anti-pattern to "headless services" - I think I may have seen references to "cardinal services".

For example, when I've set up my Kafka nodes, I find myself specifying zk1:2181,zk2:2181,zk3:2181 in config files. To the point where we're now using consul-template to dynamically generate config files.

Perhaps something like: instance[myinstanceid].mysvc.myns.service.kubernetes.local though that still leaves out how to find the (potentially) dynamic port number....

**Comment by bgrant0607** [comment]
Re. Zookeeper, see #260 (nominal services, which was formerly referred to as cardinal services), which is out of scope for now.

**Comment by thockin** [comment]
@smarterclayton I think 1b is all but decided.  The root DNS domain is already a configurable, it just happens to default to kubernetes.local.  I can easily imagine tweaking it to include cluster names:  e.g. xy.kube.local and ab.kube.local - then I could cross-ref names like mysvc.myns.svc.ab.kube.local.

Can you explain what you want wildcards to produce?

**Comment by smarterclayton** [comment]
Wildcards should resolve the same as the service (the canonical A query for the service).  It's really there to let an app easily support wildcarding for purposes of their own - security (cookie hoisting and isolation), vanity urls (myfoo.service.bar.local), and routing (I can set up a security router at *.foo.bar.service.local that proxies to other services via a mapping of * to service backend).

I was proposing not having the root DNS name be configurable - so that people can write templates and examples that work the same everywhere.  Cluster name is interesting for crossing, but I think there are advantages for consistent inside cluster DNS names.

----- Original Message -----

**Comment by thockin** [comment]
On May 6, 2015 10:14 AM, "Clayton Coleman" <notifications@github.com> wrote:
>
> Wildcards should resolve the same as the service (the canonical A query
for the service). It's really there to let an app easily support
wildcarding for purposes of their own - security (cookie hoisting and
isolation), vanity urls (myfoo.service.bar.local), and routing (I can set
up a security router at *.foo.bar.service.local that proxies to other
services via a mapping of * to service backend).
>
> I was proposing not having the root DNS name be configurable - so that
people can write templates and examples that work the same everywhere.
Cluster name is interesting for crossing, but I think there are advantages
for consistent inside cluster DNS names.

If the root is not configurable, do we want it to be kubernetes.local? I
thought we were trying to never let apps know they were in kubernetes? The
truth is that apps should almost never need to specify the full DNS name.
Relative names should work.

We should if we don't insert a level for cluster now it will hurt more
later. @ vishh

>
>
> ----- Original Message -----
> > @smarterclayton I think 1b is all but decided. The root DNS domain is
> > already a configurable, it just happens to default to kubernetes.local.
I
> > can easily imagine tweaking it to include cluster names: e.g.
xy.kube.local

**Comment by smarterclayton** [comment]
> On May 8, 2015, at 10:56 PM, Tim Hockin <notifications@github.com> wrote:
> 
> On May 6, 2015 10:14 AM, "Clayton Coleman" <notifications@github.com> wrote:
> >
> > Wildcards should resolve the same as the service (the canonical A query
> for the service). It's really there to let an app easily support
> wildcarding for purposes of their own - security (cookie hoisting and
> isolation), vanity urls (myfoo.service.bar.local), and routing (I can set
> up a security router at *.foo.bar.service.local that proxies to other
> services via a mapping of * to service backend).
> >
> > I was proposing not having the root DNS name be configurable - so that
> people can write templates and examples that work the same everywhere.
> Cluster name is interesting for crossing, but I think there are advantages
> for consistent inside cluster DNS names.
> 
> If the root is not configurable, do we want it to be kubernetes.local? I
> thought we were trying to never let apps know they were in kubernetes?

I think there's an advantage to "full name looks like any old completely isolated sandbox" but don't feel strongly.  I'm ok with ".k8s.local" or the longer "kubernetes.local" or "cluster.local"

**Comment by vishh** [comment]
Moving the discussion in #8010 to over here:

Currently namespaces are DNS names which are incompatible with the skydns' storage schema. While we are at re-working out current DNS design, I would like to understand if it is necessary for namespaces to be DNS names. Can we switch namespaces to be labels?
Example: A service with namespace `kubernetes.default` and name `foo` will result in our dns framework not creating a dns entry.
It is possible to handle this issue entirely in the DNS framework, but I want to understand some of the existing assumptions before jumping to solutions.

(@bgrant0607 and @thockin mentioned that namespaces could be labels in #8010)

Also what's the current plan for inter-cluster service discovery? Are we expecting our skydns implementation to handle that use case?

**Comment by smarterclayton** [comment]
> On May 9, 2015, at 1:50 PM, Vish Kannan <notifications@github.com> wrote:
> 
> Moving the discussion in #8010 to over here:
> 
> Currently namespaces are DNS names which are incompatible with the skydns' storage schema. While we are at re-working out current DNS design, I would like to understand if it is necessary for namespaces to be DNS names. Can we switch namespaces to be labels?
> Example: A service with namespace kubernetes.default and name foo will result in our dns framework not creating a dns entry.
> It is possible to handle this issue entirely in the DNS framework, but I want to understand some of the existing assumptions before jumping to solutions.
> 
The new plan for DNS should fix this - in Openshift (where I did most of the prototyping for the DNS backed by the service API) I believe we don't have this problem due to the way we introduce names.  So temporary problem at worst.

There may be advantages to leaving namespaces capable of themselves being sub domains, although I don't have a concrete use case right now
> (@bgrant0607 and @thockin mentioned that namespaces could be labels in #8010)
> 
> Also what's the current plan for inter-cluster service discovery? Are we expecting our skydns implementation to handle that use case?
> 
I would prefer if other clusters were visible via a subset of the local DNS namespace in a predictable way.  Although it might make more sense to more formally define the boundaries between clusters due to differences in network, security, and proxy concerns between clusters.  I expect lots of cases involving clusters in different security zones, and would want to control that via boundary firewalls or proxies.

**Comment by thockin** [comment]
The problem with namespace names being more than just DNS labels is
fundamentally that it gets ambiguous.  Services happen to be a DNS label
also (though I hoped to move that back to full subdomains).

Given a namespace "quux" hosting a service (or any other DNS-able entity -
pods?) named "foo.bar" and another namespace "bar.quux" hosting a service
named "foo", what does a DNS lookup for foo.bar.quux do?

I think it is simpler to cut this off entirely.  One or the other must be a
DNS label, and not subdomain.  Given that Namespace is common to every
object, it feels like the biggest bang for the buck.  Is there a real
rationale for keeping namespaces as DNS subdomains?

**Comment by alml** [comment]
Given that one of the design goals of namespaces is policy enforcement on per-namespace basis, it might be handy to setup access policy for "foo" project, and then each instance of this project (foo/prod, foo/test, foo/staging, etc) can land in its own namespace, but still inherit the policy.

**Comment by smarterclayton** [comment]
I have no concrete argument at the present.  If we do it, let's do it now.

**Comment by thockin** [comment]
We don't have (and I don't think we will any time soon) sub-namespaces. We
could do this through convention, though. I agree it would be better if we
could have subdomains for namespace. Is it worth making everything else be
labels instead? Something has to bend here. What's more useful. A namespace
"foo.bar" or a pod/service "foo.bar"?

I suppose that, if namespace are sufficiently descriptive everything else
is named relative to a namespace...

**Comment by smarterclayton** [comment]
So one perspective on namespaces we may want to discuss in this context (as Brian has noted in other threads) is that a namespace is a viewport through which you view a set of resources that are uniquely identified in that space by a name.  

The set of namespaces you can view (across clusters) is your working set - those namespaces might be in different datacenters, different cloud providers, be running different versions of Kube, and have different active users (this is the heterogenous view of Kube deployments).  I think the (namespace, user, server) tuple is the unit of context, and the list of those tuples is a clients view of the world.  The user/client probably cares less about the namespace being single or multilevel than it does about being able to identify one namespace for user in another - making it so that namespace A on server X can talk to B on cluster Z on cloud S.  But the namespace will probably carry a greater burden of identification in that mental model, so namespace "prod.projectfoo" seems more likely than wanting to have service "mysql.prod".

Mixing dev/stg/prod in a single namespace is probably for small teams with homogenous access, rather than massive teams with multiple acl tiers.

**Comment by thockin** [comment]
A couple things to keep this moving forward.

1) I propose to convert namespaces to DNS labels (no dots) while we sort
out what to do

2) Re-evaluate namespaces as subdomains when ready

3) Evaluate feasibility of changing all top-level objects to DNS labels
before 1.0 (may not fly)

4) I am not concinved that embedding cluster into the namespace name (if I
understand) is the right way to handle multi-cluster rigs.  we should have
that discussion on its own.  If we need to inject another layer of DNS
scoping after 1.0 we can, and I think it will hurt less.  But better to do
it now if we can.

**Comment by smarterclayton** [comment]
----- Original Message -----
> A couple things to keep this moving forward.
> 
> 1) I propose to convert namespaces to DNS labels (no dots) while we sort
> out what to do

Agreed

> 
> 2) Re-evaluate namespaces as subdomains when ready

Agreed

> 
> 3) Evaluate feasibility of changing all top-level objects to DNS labels
> before 1.0 (may not fly)

Pods being the risky one

> 
> 4) I am not concinved that embedding cluster into the namespace name (if I
> understand) is the right way to handle multi-cluster rigs.  we should have
> that discussion on its own.  If we need to inject another layer of DNS
> scoping after 1.0 we can, and I think it will hurt less.  But better to do
> it now if we can.

I'm not convinced either.

**Comment by bgrant0607** [comment]
I'm fine with changing namespaces to DNS labels. Not as many users are using namespaces, and there's a strong reason for the change.

Cluster shouldn't go in the namespace names.

I'm much less convinced we should change all other objects. That would definitely break some users, and I believe some of our own examples/configs. I'm concerned about the length restriction as much as dots.

**Comment by vishh** [comment]
Summary: Namespaces will no longer be allowed to have "." characters. The next release of kubernetes will disallow creating namespaces with '.' characters. This change will be communicated to users via 'kubernetes-announce@' and 'kubernetes-dev@' shortly.

**Comment by derekwaynecarr** [comment]
I am good with a namespace name being a DNS label as well.

**Comment by smarterclayton** [comment]
On the topic of default suffix, I think "*.cluster.local" is strictly superior to kubernetes.local.  It properly identifies scope (a cluster), is generally descriptive (the local clusters DNS scope). We may not make it required, but it's a nice generic domain, and it has the advantage of separating remote DNS entries injected into the local environment in the future (which one is my cluster?  Oh, the one called cluster.  Which one is remote?)

Opinions

**Comment by thockin** [comment]
Do you envision *.xy.cluster.local, where xy is a cluster name or do you
think .cluster.local is a fill-in for .xy.local?

I'm fine with s/kubernetes/cluster - since we'll be eventually breaking
people who use full-name queries anyway by introducing a level of nesting,
we might as well break things now.

**Comment by derekwaynecarr** [comment]
+1 for cluster as well.

**Comment by a-robinson** [comment]
It was changed to cluster.local in #8105

**Comment by derekwaynecarr** [comment]
@a-robinson thanks, i was catching up on github notifications, looks like i am 24 hrs behind ;-)

**Comment by thockin** [comment]
A minified form of our supported schema:

```
#!/bin/bash

function d {
  docker exec -ti $(docker ps | grep etcd | cut -f1 -d' ') "$@"
}

ETCDCTL=/usr/local/bin/etcdctl
echo multiport headless service with DNSRR
d $ETCDCTL rm --recursive /skydns/local/cluster/ns

# Given a headless Service 'svc', with these Endpoints:
# {
#     "subsets": [
#         {
#            "addresses": [
#                 { "ip": "1.1.1.1", "targetRef": { name: "pod1" } },
#                 { "ip": "2.2.2.2", "targetRef": { name: "pod2" } }
#             ],
#             "ports": [
#                 { "name": "http",  "port": 80,  "protocol": "TCP" },
#                 { "name": "https", "port": 443, "protocol": "TCP" }
#             ]
#         }, {
#             "addresses": [
#                 { "ip": "3.3.3.3", "targetRef": { name: "pod3" } }
#             ],
#             "ports": [
#                 { "name": "http",  "port": 8080, "protocol": "TCP" },
#                 { "name": "https", "port": 8443, "protocol": "TCP" }
#             ]
#         }
#     ]
# }

# Create records for the hosts, based on port names.  SkyDNS will infer a base name and produce A and SRV records.
# Generate a random name for the unnamed port
echo "Create per-pod records"
POD1=$RANDOM
POD2=$RANDOM
POD3=$RANDOM
d $ETCDCTL set /skydns/local/cluster/ns/svc/$POD1 '{"host":"1.1.1.1", "port":0}'
d $ETCDCTL set /skydns/local/cluster/ns/svc/$POD2 '{"host":"2.2.2.2", "port":0}'
d $ETCDCTL set /skydns/local/cluster/ns/svc/$POD3 '{"host":"3.3.3.3", "port":0}'
echo

# Create explicit SRV respords for the RFC-style service names.  SkyDNS does not recurse through records with leading underscores.
echo "Create per-pod, per-port records"
d $ETCDCTL set /skydns/local/cluster/ns/svc/_tcp/_http/$POD1 "{\"host\":\"$POD1.svc.ns.cluster.local\", \"port\":80}"
d $ETCDCTL set /skydns/local/cluster/ns/svc/_tcp/_https/$POD1 "{\"host\":\"$POD1.svc.ns.cluster.local\", \"port\":443}"
d $ETCDCTL set /skydns/local/cluster/ns/svc/_tcp/_http/$POD2 "{\"host\":\"$POD2.svc.ns.cluster.local\", \"port\":80}"
d $ETCDCTL set /skydns/local/cluster/ns/svc/_tcp/_https/$POD2 "{\"host\":\"$POD2.svc.ns.cluster.local\", \"port\":443}"
d $ETCDCTL set /skydns/local/cluster/ns/svc/_tcp/_http/$POD3 "{\"host\":\"$POD3.svc.ns.cluster.local\", \"port\":8080}"
d $ETCDCTL set /skydns/local/cluster/ns/svc/_tcp/_https/$POD3 "{\"host\":\"$POD3.svc.ns.cluster.local\", \"port\":8443}"
echo

echo "Expect multiple A records - one for each pod"
dig @10.0.0.10 +noall +answer +additional +ignore svc.ns.cluster.local A
echo

echo "Expect one SRV record for each pod-port"
dig @10.0.0.10 +noall +answer +additional +ignore _http._tcp.svc.ns.cluster.local SRV
dig @10.0.0.10 +noall +answer +additional +ignore _https._tcp.svc.ns.cluster.local SRV
echo
```

yields

```
multiport headless service with DNSRR
Create per-pod records
{"host":"1.1.1.1", "port":0}
{"host":"2.2.2.2", "port":0}
{"host":"3.3.3.3", "port":0}

Create per-pod, per-port records
{"host":"6816.svc.ns.cluster.local", "port":80}
{"host":"6816.svc.ns.cluster.local", "port":443}
{"host":"1198.svc.ns.cluster.local", "port":80}
{"host":"1198.svc.ns.cluster.local", "port":443}
{"host":"18221.svc.ns.cluster.local", "port":8080}
{"host":"18221.svc.ns.cluster.local", "port":8443}

Expect multiple A records - one for each pod
svc.ns.cluster.local.	3600	IN	A	2.2.2.2
svc.ns.cluster.local.	3600	IN	A	3.3.3.3
svc.ns.cluster.local.	3600	IN	A	1.1.1.1

Expect one SRV record for each pod-port
_http._tcp.svc.ns.cluster.local. 3600 IN SRV	10 33 80 6816.svc.ns.cluster.local.
_http._tcp.svc.ns.cluster.local. 3600 IN SRV	10 33 80 1198.svc.ns.cluster.local.
_http._tcp.svc.ns.cluster.local. 3600 IN SRV	10 33 8080 18221.svc.ns.cluster.local.
_https._tcp.svc.ns.cluster.local. 3600 IN SRV	10 33 443 6816.svc.ns.cluster.local.
_https._tcp.svc.ns.cluster.local. 3600 IN SRV	10 33 443 1198.svc.ns.cluster.local.
_https._tcp.svc.ns.cluster.local. 3600 IN SRV	10 33 8443 18221.svc.ns.cluster.local.
```

@bgrant suggested inserting subset structure here, but I am not convinced it is helpful nor that it is not harmful.  Suggest we do simplest thing now and re-evaluate that (should not affect users).

The biggest problem is with the random names - when kube2sky restarts it will need to either blow everything away (bad because DNS could fail), figure out the names that it had previously generated for equivalent pods, or else add new records then remove old records.

**Comment by bgrant0607** [comment]
Specifically, I suggested that DNS reflect the substructure we have in Endpoints, by creating one SRV record and one CNAME per subset, with each CNAME expanding into A records for all pods in that subset. In addition to whatever efficiency might be gained from fewer records and fewer lookups, it would also make it easier for a human to understand the topology of the service (e.g., how many pods used each port combination).

However, I'm fine with the above proposal for now. Either way, multiple SRV, CNAME, and A records are required for a service with multiple subsets.

**Comment by thockin** [comment]
I think this is done!
